filament/src/MorphTargets.cpp

```diff
@@ -0,0 +1,219 @@
+    mImpl->mVertexCount = vertexCount;
+    return *this;
+}
+
+MorphTargets::Builder& MorphTargets::Builder::count(size_t count) noexcept {
+    mImpl->mCount = count;
+    return *this;
+}
+
+MorphTargets* MorphTargets::Builder::build(Engine& engine) {
+    return upcast(engine).createMorphTargets(*this);
+}
+
+// ------------------------------------------------------------------------------------------------
+
+constexpr size_t MAX_TEXTURE_WIDTH = 4096;
+
+static const auto FREE_CALLBACK = [](void* mem, size_t, void*) { free(mem); };
+
+static inline size_t getWidth(size_t vertexCount) noexcept {
+    return std::min(vertexCount, MAX_TEXTURE_WIDTH);
+}
+
+static inline size_t getHeight(size_t vertexCount) noexcept {
+    return (vertexCount + MAX_TEXTURE_WIDTH) / MAX_TEXTURE_WIDTH;
+}
+
+static inline size_t getDepth(size_t targetCount) noexcept {
+    return targetCount * 2;
+}
+
+static inline size_t getSize(size_t vertexCount) noexcept {
+    const size_t stride = getWidth(vertexCount) * sizeof(float4);
+    return Texture::PixelBufferDescriptor::computeDataSize(
+            Texture::PixelBufferDescriptor::PixelDataFormat::RGBA,
+            Texture::PixelBufferDescriptor::PixelDataType::FLOAT,
+            stride, getHeight(vertexCount), 1);
+}
+
+static inline size_t getPositionIndex(size_t targetIndex) noexcept {
+    return targetIndex * 2 + 0;
+}
+
+static inline size_t getTangentIndex(size_t targetIndex) noexcept {
+    return targetIndex * 2 + 1;
+}
+
+FMorphTargets::FMorphTargets(FEngine& engine, const Builder& builder)
+        : mSBuffer(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT),
+          mVertexCount(builder->mVertexCount),
+          mCount(builder->mCount) {
+    FEngine::DriverApi& driver = engine.getDriverApi();
+
+    mSbHandle = driver.createSamplerGroup(PerRenderPrimitiveMorphingSib::SAMPLER_COUNT);
+
+    mTbHandle = driver.createTexture(SamplerType::SAMPLER_2D_ARRAY, 1, TextureFormat::RGBA32F, 1,
+            getWidth(mVertexCount), getHeight(mVertexCount), getDepth(mCount),
+            TextureUsage::DEFAULT);
+
+    mSBuffer.setSampler(PerRenderPrimitiveMorphingSib::TARGETS,
+            { mTbHandle, { .filterMin = SamplerMinFilter::NEAREST, .filterMag = SamplerMagFilter::NEAREST } });
+}
+
+void FMorphTargets::terminate(FEngine& engine) {
+    FEngine::DriverApi& driver = engine.getDriverApi();
+
+    driver.destroySamplerGroup(mSbHandle);
+    driver.destroyTexture(mTbHandle);
+}
+
+void FMorphTargets::setPositionsAt(FEngine& engine, size_t targetIndex, math::float3 const* positions, size_t count) {
+    ASSERT_PRECONDITION(targetIndex < mCount, "targetIndex must be < count");
+
+    auto size = getSize(mVertexCount);
+
+    ASSERT_PRECONDITION((int)sizeof(math::float3) * count <= size,
+            "MorphTargets (size=%lu) overflow (size=%lu)",
+            size, sizeof(math::float3) * count);
+
+    auto* out = (float4*) malloc(size);
```

Comment:
Maybe add a comment here that we could use a pool instead of `malloc()` directly.

Reply:
Ok it's done.